feat: Responsive area line chart

maloun96 opened the pull request:
Rotate labels if they are overlapping.

https://user-images.githubusercontent.com/8277264/110635316-a0182c80-81b3-11eb-9a1b-66eafd45f011.mov


https://user-images.githubusercontent.com/8277264/110805893-3454c300-828a-11eb-89c4-b9c2a6eb26be.mov

maloun96 committed:
fix: add local_prefix for table chart

maloun96 committed:
Rotate labels if they overlap, hide legend if it overlaps with chart. Hide xAxis if labels are rotated and still overlap

add callback on renderEnd

ResizableChartDemo emit event on resizeStop

junlincc reviewed (commented):
Thanks for the PR, Victor! LGTM. 
truncating x-axis label on resize is one of the requirements as we show date in tooltip. 
are you planning to address this in a separate PR?

maloun96 commented:
@rusackas @villebro @ktmud Let's start the discussion. So for the nvd3 chart to work properly I added the overlapping check function, and basically, I moved all the NVD code to a separate file, and when the chart is rendered it will check the overlapping legend/labels and will rerender again. I tested for some performance and it's not a big difference.
What do you think?

junlincc commented:
@maloun96 
In general, when a chart is resized, with the limit space, legend should go(hide) first, then the x-axis label, and y-axis label, we need to make sure the chart always shows on resizing, user can still see details on tooltip. 

One thing i notice is that the dates x-axis label(in certain format) take up a lot of room vertically, so we can either hide them completely or truncate them.

Ideally, 1) if the x-axis labels are set horizontal, they should rotate once overlap. 
2) the space x-axis label take should be limited to make sure chart always show.

It sounds complex to me lol. Echarts seem to have pretty good out-of-the box solution. @villebro what's your take on further invest in the responsiveness in NVD3 vs. go straight to Echarts?

maloun96 commented:
so for now we check if labels are overlapping then we rotate it to 45 deg and if the legend is overlapping with the chart also will be hidden and if labels are rotated and still overlapping then we will hide xlabel


https://user-images.githubusercontent.com/8277264/110853690-99c1a780-82bc-11eb-9d34-0bb4ffb1c936.mov


https://user-images.githubusercontent.com/8277264/110853729-a47c3c80-82bc-11eb-99d6-2d29dff8dc41.mov

maloun96 commented:
https://user-images.githubusercontent.com/8277264/111117354-16c87780-8570-11eb-8812-590e1e8091a6.mov


https://user-images.githubusercontent.com/8277264/111117374-1b8d2b80-8570-11eb-98f9-44821d70d141.mov


https://user-images.githubusercontent.com/8277264/111117380-1d56ef00-8570-11eb-955b-85ab30dbdff2.mov


https://user-images.githubusercontent.com/8277264/111117389-1fb94900-8570-11eb-9f23-027d9c08fd99.mov


https://user-images.githubusercontent.com/8277264/111117400-22b43980-8570-11eb-8ba3-35eba734da45.mov


https://user-images.githubusercontent.com/8277264/111117403-23e56680-8570-11eb-90b9-360526c99feb.mov

maloun96 committed:
Refactor area chart

ktmud commented:
I remember x-axis label rotation is customizable for some charts. Does this only apply to those with the default rotation?

villebro commented:
The codebase on this repo has been moved to the main [Apache Superset repo](https://github.com/apache/superset), and consequently the repo is in the process of being archived. See the Superset Improvement Proposal for details: https://github.com/apache/superset/issues/13013 . While all currently open issues and PRs will be closed, we encourage you to reopen this PR on the main repo, which should be as simple as moving over any code changes as follows:
- core packages: `packages/**` (this repo) -> `superset-frontend/packages/**` (main repo)
- plugins: `plugins/**` (this repo) -> `superset-frontend/plugins/**` (main repo)

If you need help with the migration, please post a message on the SIP or reach out on the [community Slack](https://join.slack.com/t/apache-superset/shared_invite/zt-uxbh5g36-AISUtHbzOXcu0BIj7kgUaw).
